x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameIndexer.java

```diff
@@ -226,6 +293,72 @@ protected boolean handleCircuitBreakingException(Exception e) {
         return true;
     }
 
+    private void getChangedBuckets(DataFrameTransformCheckpoint oldCheckpoint, DataFrameTransformCheckpoint newCheckpoint,
+            ActionListener<Map<String, List<String>>> listener) {
+        SearchRequest searchRequest = new SearchRequest(getConfig().getSource().getIndex());
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+
+        // we do not need the sub-aggs
+        CompositeAggregationBuilder changesAgg = pivot.buildChangedBucketsAggregation(pageSize);
+        sourceBuilder.aggregation(changesAgg);
+        sourceBuilder.size(0);
+
+        QueryBuilder pivotQueryBuilder = getConfig().getSource().getQueryConfig().getQuery();
+
+        DataFrameTransformConfig config = getConfig();
+        if (config.getSyncConfig() != null) {
+            BoolQueryBuilder filteredQuery = new BoolQueryBuilder().
+                    filter(pivotQueryBuilder).
+                    filter(config.getSyncConfig().getRangeQuery(oldCheckpoint, newCheckpoint));
+
+            logger.trace("Gathering changes using query {}", filteredQuery);
+            sourceBuilder.query(filteredQuery);
+        } else {
+            logger.trace("No sync configured");
+            listener.onResponse(null);
+            return;
+        }
+
+        searchRequest.source(sourceBuilder);
+        searchRequest.allowPartialSearchResults(false);
+
+        // pre-fill the hashmap with the expected keys
+        // note that changed bucket might be empty due to the query
+        HashMap<String, List<String>> keys = new HashMap<>();
+        for(Entry<String, SingleGroupSource> entry: config.getPivotConfig().getGroupConfig().getGroups().entrySet()) {
```

Comment:
Similar to above, I think a non-trivial optimization would be to only do `terms`